Xamarin/DrawXShared/RealmDraw.cs

```diff
@@ -0,0 +1,552 @@
+    - A long, winding path being updated - there may be enough lag that we have to add more than 
+      one point to it locally but we don't want to redraw the whole thing from scratch.      
+    - Many small, single "dab" strokes being drawn, say from someone tapping a display, 
+      which mean we have, at least, a TouchesBegan and TouchesEnded and probably AddPoint in between.
+    
+    We make simply monitor changes - any *added* or *changed* paths get redrawn. 
+
+    Most importantly, to get the fastest possible response as the user moves their finger,
+    we draw the local line immediately as a continuation of the path they started drawing 
+    earlier. (See _currentlyDrawing)
+    
+    */
+    public class RealmDraw
+    {
+        private const float NORMALISE_TO = 4000.0f;
+        private const float PENCIL_MARGIN = 4.0f;
+        private const float INVALID_LAST_COORD = -1.0f;
+        private float _lastX = INVALID_LAST_COORD;
+        private float _lastY = INVALID_LAST_COORD;
+
+        #region Synchronised data
+        private Realm _realm;
+        private IQueryable<DrawPath> _allPaths;  // we observe all and filter based on changes
+        #endregion
+
+        #region GUI Callbacks
+        internal Action RefreshOnRealmUpdate { get; set; }
+
+        internal Action CredentialsEditor { get; set; }
+        #endregion
+
+        #region DrawingState
+        private bool _isDrawing = false;
+        private bool _ignoringTouches = false;
+        private DrawPath _drawPath;
+        private SKPath _currentlyDrawing;  // caches for responsive drawing on this device
+        private float _canvasWidth, _canvasHeight;
+        private IList<DrawPath> _pathsToDraw = null;  // set in notification callback
+        #endregion
+
+        #region CachedCanvas
+        private int _canvasSaveCount;  // from SaveLayer
+        private bool _hasSavedBitmap = false;  // separate flag so we don't rely on any given value in _canvasSaveCount
+        private bool _redrawPathsAtNextDraw = true;
+        #endregion
+
+        #region Touch Areas
+        private SKRect _loginIconRect;
+        private SKRect _loginIconTouchRect;
+        //// setup in DrawBackground
+        private float _pencilWidth;
+        private float _pencilsTop;
+        private int _numPencils;
+        private List<SKBitmap> _pencilBitmaps;
+        private SKBitmap _loginIconBitmap;
+        #endregion
+
+        #region LoginState
+        private bool _waitingForLogin = false;
+        #endregion
+
+        #region Settings
+        private DrawXSettings Settings => DrawXSettingsManager.Settings;
+
+        private int _currentColorIndex = -1;  //// for quick check if pencil we draw is current color
+
+        private SwatchColor _currentColor;
+
+        private SwatchColor CurrentColor
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(_currentColor.Name))
+                {
+                    InitCurrentColor();
```

Comment:
Couldn't there just be a default initial color? Also, why do we need to state variables for the active color?